rust_src/src/webrender_backend/cursor.rs

```diff
@@ -0,0 +1,120 @@
+use super::output::OutputRef;
+
+use lisp::{
+    remacs_sys::{
+        draw_glyphs_face, draw_phys_cursor_glyph, get_phys_cursor_geometry, get_phys_cursor_glyph,
+        glyph_row, glyph_type,
+    },
+    window::LispWindowRef,
+};
+
+pub fn draw_filled_cursor(mut window: LispWindowRef, row: *mut glyph_row) {
+    unsafe { draw_phys_cursor_glyph(window.as_mut(), row, draw_glyphs_face::DRAW_CURSOR) };
+}
+
+pub fn draw_hollow_box_cursor(mut window: LispWindowRef, row: *mut glyph_row) {
+    let cursor_glyph = unsafe { get_phys_cursor_glyph(window.as_mut()) };
+
+    if cursor_glyph.is_null() {
+        return;
+    }
+
+    let mut x: i32 = 0;
+    let mut y: i32 = 0;
+    let mut height: i32 = 0;
+    unsafe {
+        get_phys_cursor_geometry(
+            window.as_mut(),
+            row,
+            cursor_glyph,
+            &mut x,
+            &mut y,
+            &mut height,
+        )
+    };
+    let width = window.phys_cursor_width;
+
+    let frame = window.get_frame();
+    let output: OutputRef = unsafe { frame.output_data.wr.into() };
+
+    output.canvas().draw_hollow_box_cursor(x, y, width, height);
+}
+
+pub fn draw_bar_cursor(
+    mut window: LispWindowRef,
+    row: *mut glyph_row,
+    cursor_width: i32,
+    is_hbar: bool,
+) {
+    let frame = window.get_frame();
+    let output: OutputRef = unsafe { frame.output_data.wr.into() };
+
+    let cursor_glyph = unsafe { get_phys_cursor_glyph(window.as_mut()) };
+
+    if cursor_glyph.is_null() {
+        return;
+    }
+
+    if unsafe {
+        (*cursor_glyph).type_() == glyph_type::XWIDGET_GLYPH
+            || (*cursor_glyph).type_() == glyph_type::IMAGE_GLYPH
+    } {
+        return;
+    }
+
+    let face = unsafe {
+        let face_id = (*cursor_glyph).face_id();
+        let face_id = std::mem::transmute::<u32, lisp::remacs_sys::face_id>(face_id);
```

Comment:
Is there anyway to avoid mem::transmute here?

Reply:
No, face_id is an enum. We can only convert u32 to enum by mem::transmute.